### PR TITLE
Fix freezes due to large circular arc radius

### LINKED ISSUE #5538: Dragging the end anchor point to the start of the slider causes a crash
**Describe the crash:** 
When youre in the editor, selecting a slider gives you the option to change the angle of them with those anchor points ( i hope they were called that way ) when you drag the end to the start the slider just expands and the game crashes 

**Screenshots or videos showing encountered issue:** 
https://streamable.com/je2eo

**osu!lazer version:** 
8caf9ad 

**Logs:** 
[performance.log](https://github.com/ppy/osu/files/3448579/performance.log)
[runtime.log](https://github.com/ppy/osu/files/3448580/runtime.log)

**Computer Specifications:** 
i3 6100 8GB ram GTX 1050Ti Windows 10 x64 1903

### PR DESCRIPTION
**Pull request by Naxesss**
(This is my first PR here so do let me know if I'm doing anything incorrectly.)

This resolves https://github.com/ppy/osu/issues/5538 and the freezing/crashing portion of https://github.com/ppy/osu/issues/12047. @bdach outlines this issue pretty well in both of those.

We could potentially fix this in [`PathApproximator`'s `ApproximateCircularArc` in the osu!framework](https://github.com/ppy/osu-framework/blob/master/osu.Framework/Utils/PathApproximator.cs#L165) instead, but this poses the risk of arbitrarily limiting the radius of circular arcs based on osu's playfield dimensions, so I think we should solve this in the client.

It is also in the client where the performance bottleneck seems to occur; placing a breakpoint after the `PathApproximator.ApproximateCircularArc` line revealed that it's hit almost instantly (compared to how long the freezes last) even for cases with 4k+ subpath points. I noticed split-second freezing after ~500 points in-game, so 400 should give a decent margin while still allowing large circles to be made.

I am a bit unsure of where to put the test. I was considering splitting it off into a `TestSceneSliderPlacementEdgeCaseBlueprint` class, as the tests within `TestSceneSliderPlacementBlueprint` are mostly regular use cases. However, if I do that then the `assertLength` etc functions would need to be duplicated unless moved to another class and inherited by both. I figured it'd be worth asking about that before possibly wasting effort if the current is ok.

- [x] depends on https://github.com/ppy/osu-framework/pull/4313

### CODE REVIEW

**Comment by bdach** [comment]
A few things:

> We could potentially fix this in `PathApproximator`'s `ApproximateCircularArc` in the osu!framework instead

I agree with not doing this, and it was also the consensus from previous discussions on how to approach this that this should be a game-local thing.

> It is also in the client where the performance bottleneck seems to occur

To clarify this, the bottleneck is not computing the slider path, it's drawing it. For the slider to be drawn, FBOs are allocated over at the framework side; in the regressive cases the radius of the slider increases by a lot, causing frame drops because of the sheer amount of time required to render it.

---

While predicating the cutoff on the number of points in the approximation is certainly a way to check whether the slider is too large, it's not exactly what stable does. For reference, stable does this:

```csharp
                    OsuMathHelper.CircleThroughPoints(a, b, c, out centre, out radius, out t_initial, out t_final);

                    if (GameBase.Mode == OsuModes.Edit)
                    {
                        // Find the circle's bounding box
                        // Don't fall back to bezier if the slider is already placed and
                        // eg. the user rotates it.
                        // https://osu.ppy.sh/forum/t/116612
                        osu.Helpers.BoundingBox box = new Helpers.BoundingBox(OsuMathHelper.CirclePoint(centre, radius, t_initial), OsuMathHelper.CirclePoint(centre, radius, t_final));
                        double t_low = Math.Min(t_initial, t_final);
                        double t_high = Math.Max(t_initial, t_final);
                        int q = (int)(t_low * 2.0d / Math.PI + 1);
                        while (q * 0.5f * Math.PI < t_high)
                        {
                            box.Add(OsuMathHelper.CirclePoint(centre, radius, q * 0.5f * Math.PI));
                            q++;
                        }

                        if ((placingPending && ((float)box.Width > 512 || (float)box.Height > 384))
                            || ((float)box.Width > 640 || (float)box.Height > 640))
                            goto case CurveTypes.Bezier;
                    }
```

In other words, this Bezier conversion logic is only active in the editor. The change as done here would affect all sliders, including the ones in gameplay, therefore possibly breaking maps that bypassed that editor limitation (via manual editing or otherwise), so I'm not sure we should proceed with this fix.

I had looked at this recently, and attempted to make this logic local to the editor by placing that bounding box limiting logic in `PathControlPointPiece` if I recall correctly, but it was not working all that well due to reasons that I haven't managed to look into too much yet.

**Comment by Naxesss** [comment]
> In other words, this Bezier conversion logic is only active in the editor, and only in specific cases. The change as done here would affect all sliders, including the ones in gameplay, therefore possibly breaking maps that bypassed that editor limitation (via manual editing or otherwise), so I'm not sure we should proceed with this fix.

That is a good point, I didn't realize it was limited to the editor. In that case this specific fix could possibly break aspire maps, so will close this.

**Comment by peppy** [comment]
We might still be able to use an approach like this but apply it closer to the editor's code. I haven't looked into it in detail, but for instance it may be possible to replace the `SliderPath` with a custom `EditSliderPath`, or something similar via `DrawableOsuEditRuleset`.

Either way, I'd probably keep this PR open for the time being until we decide that it's 100% not a valid direction to follow.

**Comment by bdach** [comment]
`SliderPath` is a hitobject-level entity, not a drawable entity. So it'd probably have to be done at a deeper level than at the drawable ruleset (at conversion time or something?)

**Comment by peppy** [comment]
Even there, we so have control over this via `SliderPlacementBlueprint`, but that does seem like it'd bee too deep.

Another option may be running the calculation once locally in `PathControlPointVisualiser` before actually applying the changes to the `Slider` itself, to catch the excessive case.

**Comment by frenzibyte** [comment]
Also, as an aside, do not open pull requests off of a `master` branch, give the [contributing guidelines](https://github.com/ppy/osu/blob/master/CONTRIBUTING.md#i-would-like-to-submit-a-pull-request) a read if you haven't already.

**Comment by Naxesss** [comment]
Alright reopening.

@frenzibyte I did read them earlier but forgot about that part. Thanks though, will keep in mind for next time (assuming I can't change branches of this PR).

I've looked around a few of the other editor-related classes and as far as I've gathered, `PathControlPointVisualizer` doesn't currently seem to be receptive to moving points, only selecting, creating, and deleting. `PathControlPointPiece` has `OnDrag`, which catches moving nodes, but not the preview when initially placing the slider. So that's a problem.

Having played around with `OnDrag` a bit, however, it seems possible to isolate the control points responsible for creating a circular arc, and then running their positions through the bezier approximation to then decide whether to temporarily change the PathType, like this (note the control point changing colour):



It's not a very elegant solution, though, as this calculates the bezier approximation twice for every path (once to know which path type to use and once in `SliderPath`) - can probably reuse/cache that somehow. Also currently, if you let go of the slider when it is in bezier mode, it stays that way (the original `PathType` is stored only for each drag), unlike in stable.

**Comment by bdach** [comment]
@Naxesss I think that gif looks like a good start, and I wouldn't really put much mind to the fact that releasing in an invalid position will convert the node to Bezier myself. Especially given it's an area in which stable would essentially silently lie to the user about what it's doing in the editor, and reproducing stable's behaviour would probably result in way too much complexity for what it's worth.

That said if we do go in this direction, it would be probably best to also disable the perfect curve type in the control point's context menu, which sounds like it will not be easy or pretty...

**Comment by peppy** [comment]
Having the colour / type change is definitely what we want, rather than hiding it behind internal switches like stable does. It should probably only "commit" the curve type when letting go of the drag, so the user can return to a sane `perfect` curve if that is their intention.

**Comment by smoogipoo** [comment]
This may be fine as a temporary solution, but I don't think it's viable in the long term. My primary concern is, with this being done at `SliderPath` level, breaking maps that take advantage of these mechanics (e.g. https://osu.ppy.sh/beatmapsets/756794#osu/1605148), so I'll at least have to run a diffcalc of these changes.

Furthermore, this is still going to wreck systems that have lower VRAM than yours:
image
So the better solution is to limit FBO sizes, but since that hasn't been done yet this temporary change is fine. Just keep in mind that these sort of scenarios can happen in-game even with your heuristic - 200MB can be quite large on some (especially integrated/mobile) systems.

As an aside, all stable does is limit the bounding box to 640x480 before changing to bezier. If that can be done instead directly in the blueprints rather than `SliderPath` itself, I'd highly prefer it.

**Comment by Naxesss** [comment]
> Maybe it is worth it to come up with a slightly different test for this scenario (e.g. one that relies on the dot product/cosine of the two vectors spanning the arc) to avoid computing the path every time?

One alternative that works well is this:
```csharp
if (pathType == PathType.PerfectCurve)
{
    Vector2[] points = pointsInSegment.Select(point => point.Position.Value).ToArray();
    if (points.Length < 3)
        return;

    Vector2 a = points[0];
    Vector2 b = points[1];
    Vector2 c = points[2];

    float maxLength = points.Max(p => p.Length);

    Vector2 normA = new Vector2(a.X / maxLength, a.Y / maxLength);
    Vector2 normB = new Vector2(b.X / maxLength, b.Y / maxLength);
    Vector2 normC = new Vector2(c.X / maxLength, c.Y / maxLength);

    float det = (normA.X - normB.X) * (normB.Y - normC.Y) - (normB.X - normC.X) * (normA.Y - normB.Y);

    float acSq = (a - c).LengthSquared;
    float abSq = (a - b).LengthSquared;
    float bcSq = (b - c).LengthSquared;

    // Exterior = curve wraps around the long way between end-points
    // Exterior bottleneck is drawing-related, interior bottleneck is approximation-related,
    // where the latter is much faster, hence differing thresholds
    bool exterior = abSq > acSq || bcSq > acSq;
    float threshold = exterior ? 0.05f : 0.001f;

    if (Math.Abs(det) < threshold)
        pointsInSegment[0].Type.Value = PathType.Bezier;
}
```
Which catches both cases of freezing:
- Middle point being beside the end-points, where as @bdach explained earlier we quickly run out of memory due to FBO allocation.
- Middle point being in between the end-points, where not much is rendered so I'd assume must be approximation-related (possibly the list of subpaths becoming very large as the radius approaches infinity).

---

I've tried various ways of applying this, and the only method I've found that works for all slider path changes (both when initially placing and when moving already placed control points), is to subscribe to control point changes in `SliderPath`. This way only *editing* is affected, while *gameplay* is kept the same. So the map @smoogipoo linked still works like in stable, for example.

However by doing it here instead of in `PathControlPointPiece`, we lose the context of the drag, which means every time it changes the path to Bezier, it "commits" as @peppy mentioned shouldn't happen. A workaround for this is to constantly try to maintain the path type while dragging/placing (where `SliderPath` updates it last), giving us this behaviour as desired:





(I did also look at the blueprint classes, but they are about placing new objects, not modifying existing ones from what I can tell. So that would at best only catch the case in the latter gif, if I'm not mistaken.)

Will push the code so we can take a look at it. I added a `PointsInSegment` method in `SliderPath`, as I noticed a very similar implementation was done by @bdach [here on another issue](https://github.com/ppy/osu/issues/11885#issuecomment-803643564).

**Comment by bdach** [comment]
The gifs look really good, but I'm not super sure still on placing that logic in `SliderPath`. Feels too close for comfort, if you will.

I'll see what everyone else thinks first, but a possible avenue that would make this nicer for me would be to have some sort of `SliderEditorManager` (bad name, just to illustrate purpose), shared by placement and selection blueprints, that would operate on a basis of something like:

```
backup slider state from before change

while change is not committed:
	check current path state
	if path invalid
		bring path to correct state (by rolling back, changing node types, etc.)

on commit, discard backed up state (because it was either re-applied, or is no longer relevant)
```

It would make for one central place for loosely enforcing the invariants that a slider path must adhere to, rather than spreading them all around the place or putting them somewhere where they may possibly impact gameplay if someone is not careful.

**Comment by peppy** [comment]
The code is pretty sound, but can agree that we *probably* don't want it in `SliderPath`, although I would rather it be there than having this issue remain broken.

I think it may be feasible to just move the latest code you have quite verbatim into `PathControlPointVisualiser`, using the bindable flow there. Give it a try and let us know if you run into any difficulties.

**Comment by Naxesss** [comment]
Alright moved to `PathControlPointVisualizer`, so should be entirely local to the editor now. Also added a few more tests (including ones for dragging control points) to ensure this sticks.

**Comment by peppy** [comment]
The implementation is looking a lot better now, but the issue mentioned [in this comment](https://github.com/ppy/osu/pull/12122#issuecomment-803965298) still stands: the limiting needs to be more stringent.

20210324 132752 (dotnet)

As it stands, I can still make a slider which uses 200mb of RAM, and double that when selecting the blueprint itself. Rather than getting to a point where the slider is 8k pixels, the limiting should kick in closer to when the slider exceeds the playfield size.

**Comment by Naxesss** [comment]
Just to be clear, would making the threshold more strict in `validCircularArc` solve this? For example upping it from 0.05 to 0.4:

**Comment by peppy** [comment]
Yep, should be fine. It should be high enough that the resultant slider is larger than the visible screen area so it doesn't falsely trigger. What you have in the screen capture seems a lot more appropriate.

**Comment by Naxesss** [comment]
Oh actually increasing that threshold might not be a good idea since it doesn't really care about size, so that would also stop smaller circles from being made... I'll try something else.

**Comment by Naxesss** [comment]
@peppy do I keep all the math logic inside `PathControlPointVisualizer`? Seems a bit strange to have so much specific math logic there, but not sure where else to put it. Figured perhaps the slider itself kept track of its bounding box but that seems to be the drawable, and unsure how/if I should access that.

**Comment by bdach** [comment]
It does feel wrong to have essentially a copy of the path approximating logic from framework in here. I see two moderately easy ways out of this:

* either to move this bounding-box computing method to framework (reusing as much of the approximator code there as possible),
* or to create a copy of the existing path, apply the user changes, and then compute the entire path (see [this old discord message I dug up](https://discord.com/channels/188630481301012481/188630652340404224/775341077481783307) from when I originally tried this and gave up for reasons I don't entirely recall).

The latter is easier because it won't require any framework changes, but it will be a lot more wasteful wrt allocations.

**Comment by FamousPig** [comment]
I stumbled across an issue. The curve type isn't reverted as perhaps expected when scaling.

https://user-images.githubusercontent.com/22643729/112827253-5de55b00-908e-11eb-9ab6-9eb463bef275.mp4

File: osu.Game.Rulesets.Osu/Edit/Blueprints/Sliders/Components/PathControlPointVisualiser.cs [hunk 1]
```diff
@@ -142,6 +148,23 @@ public void OnReleased(PlatformAction action)
...
+        private void updatePathTypes()
+        {
+            foreach (PathControlPoint segmentStartPoint in slider.Path.ControlPoints.Where(p => p.Type.Value != null))
+            {
+                if (segmentStartPoint.Type.Value != PathType.PerfectCurve)
+                    continue;
+
+                ReadOnlySpan<Vector2> points = slider.Path.PointsInSegment(segmentStartPoint).Select(p => p.Position.Value).ToArray();
+                RectangleF boundingBox = PathApproximator.CircularArcBoundingBox(points);
+                if (points.Length == 3 && boundingBox.Area >= 640 * 480)
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Edit/Blueprints/Sliders/Components/PathControlPointVisualiser.cs:163]: An area check seems a little wrong in this context. For instance, a 20x15359 slider would still theoretically pass it. It's probably harder to abuse this than that random example, given that the bounding box is based on an arc, but I'd say checking individual dimensions would be preferable.

File: osu.Game.Rulesets.Osu/Edit/Blueprints/Sliders/Components/PathControlPointVisualiser.cs [hunk 2]
```diff
@@ -91,6 +93,8 @@ private void onControlPointsChanged(object sender, NotifyCollectionChangedEventA
                         }));
 
                         Connections.Add(new PathControlPointConnectionPiece(slider, e.NewStartingIndex + i));
+
+                        point.Changed += updatePathTypes;
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Edit/Blueprints/Sliders/Components/PathControlPointVisualiser.cs:97]: This subscription seems leaky (likely holds a reference to visualiser instances after they've been disposed). Probably need to double-check if this is safe and unsubscribe from this event for all points in `Dispose()` if not.

Comment (Naxesss) [reply osu.Game.Rulesets.Osu/Edit/Blueprints/Sliders/Components/PathControlPointVisualiser.cs]: Binding to `point.Position` breaks the (probably desired) behaviour seen in the bottom gif of [this post](https://github.com/ppy/osu/pull/12122#issuecomment-804190126).

Ended up creating a local bound copy of the `Path.Version` bindable inside `PathControlPointPiece`. This way I can also remove the loop, as all control points will now call this when the path updates.

**Comment by peppy** [comment]
This is going to need some minor conflict resolution. Let me know if you're unsure of the correct merge path.

**Comment by peppy** [comment]
Looks like this doesn't guard correctly for perfect curve sections in a multi-segmented slider:

https://user-images.githubusercontent.com/191335/113860819-70c00400-97e1-11eb-98de-048b9e714478.mp4

**Comment by Naxesss** [comment]
Cause seems to be that `PointsInSegment` doesn't update when changing control point types in the slider.
Solved that by tracking the path types, using `pathTypes` to hold the bound copy references.

Also added a regression test for this case.

Before:
image

After:
image

**Review by peppy — APPROVED** [review]
This is looking and feeling pretty correct now. Good stuff!
